src/python/pants/engine/target.py

```diff
@@ -714,33 +714,51 @@ def expect_single(self) -> Target:
         return self[0]
 
 
-# If this predicate returns false, then the target's field's deps will be ignored.
-ShouldTraverseDepsPredicate: TypeAlias = (
-    "Callable[[Target, Dependencies | SpecialCasedDependencies], bool]"
-)
+@dataclass(frozen=True)
+class ShouldTraverseDepsPredicate(metaclass=ABCMeta):
+    """This callable determines whether to traverse through deps of a given Target + Field.
 
+    This is a callable dataclass instead of a function to avoid issues with hashing closures.
+    Only the id is used when hashing a function; any closure vars are NOT accounted for.
 
-class DepsTraversalPredicates:
-    """All methods on this class are ShouldTraverseDepsPredicate implementations."""
+    NB: When subclassing a dataclass (like this one), generated methods on a dataclass are NOT
+    inherited because the @dataclass decorator only inspects the current class before generating
+    a function--the decorator does NOT inspect __mro__.
+    """
 
-    @staticmethod
-    def if_dependencies_field(
-        target: Target, field: Dependencies | SpecialCasedDependencies
-    ) -> bool:
-        """This is the default ShouldTraverseDepsPredicate implementation.
+    # This _callable field ensures that __call__ is included in the __hash__ method generated by @dataclass.
+    _callable: Callable[
+        [Target, Dependencies | SpecialCasedDependencies], bool
+    ] = dataclasses.field(init=False, repr=False)
 
-        This skips resolving dependencies for fields (like SpecialCasedDependencies) that are not
-        subclasses of Dependencies.
-        """
+    def __post_init__(self):
+        object.__setattr__(self, "_callable", type(self).__call__)
+
+    @abstractmethod
+    def __call__(self, target: Target, field: Dependencies | SpecialCasedDependencies) -> bool:
+        """If this predicate returns false, then the target's field's deps will be ignored."""
+
+
+@dataclass(frozen=True)
```

Comment:
do the subclasses actually need the dataclass decorator at all?

Reply:
Without it, aren't they unfrozen? I'll try dropping the decorator on these subclasses since they don't add any fields.

Reply:
Good catch. It works just fine without decorating the subclasses. I will have to add the decorator on the predicate that stops at package targets because that will add a field. These, however, don't need it.

Reply:
I actually think we should still decorate:
- It'll work correctly when someone adds a field later
- When people copy+paste this code elsewhere they'll have it
My 2c

Reply:
I think it'll come down to preference then. I'm for not having them until needed, but won't object/block if you prefer to keep them.